module/zfs/dbuf.c

```diff
@@ -381,6 +428,41 @@ dbuf_hash_insert(dmu_buf_impl_t *db)
 	return (NULL);
 }
 
+/*
+ * This returns whether this dbuf should be stored in the metadata cache, which
+ * is based on whether it's from one of the dnode types that store data related
+ * to traversing dataset hierarchies.
+ */
+static boolean_t
+dbuf_include_in_metadata_cache(dmu_buf_impl_t *db)
+{
+	DB_DNODE_ENTER(db);
+	dmu_object_type_t type = DB_DNODE(db)->dn_type;
+	DB_DNODE_EXIT(db);
+
+	/* Check if this dbuf is one of the types we care about */
+	if (DMU_OT_IS_METADATA_CACHED(type)) {
+		/* If we hit this, then we set something up wrong in dmu_ot */
+		ASSERT(DMU_OT_IS_METADATA(type));
+
+		/*
+		 * Sanity check for small-memory systems: don't allocate too
+		 * much memory for this purpose.
+		 */
+		if (refcount_count(&dbuf_caches[DB_DBUF_METADATA_CACHE].size) >
+		    dbuf_metadata_cache_max_bytes) {
+			DBUF_STAT_BUMP(metadata_cache_overflow);
+			DTRACE_PROBE1(dbuf__metadata__cache__overflow,
+			    dmu_buf_impl_t *, db);
```

Comment:
```
/include/zfs/sys/trace.h:46:2: error: implicit declaration of function 'trace_zfs_dbuf__metadata__cache__overflow'; did you mean 'trace_zfs_dbuf__evict__one_enabled'? [-Werror=implicit-function-declaration] trace_zfs_##name((arg1))
```

You'll need to add a new tracepoint to `trace_dbuf.h` to address the kernel.org built-in failure which enables all the tracepoints.  Or alternately, we could just drop the tracepoint since you've added a kstat for this.